src/actions/upsert.ts

```diff
@@ -0,0 +1,45 @@
+import { Async, ReaderT } from "crocks";
+import { compose, ifElse, isNil, mergeRight, prop } from "ramda";
+import type { Config } from "../types.ts";
+
+//import { z } from "zod";
+interface H {
+  data: {
+    add: (doc: unknown) => typeof Async;
+    get: (id: string) => typeof Async;
+    update: (id: string, doc: unknown) => typeof Async;
+  };
+}
+
+const { of, ask, lift } = ReaderT(Async);
+
+// @ts-ignore allow any type
+const validate = (schema) =>
+  (doc: unknown) => {
+    const { success, data, errors } = schema.safeParse(doc);
+    return success
+      ? Async.Resolved(data)
+      : Async.Rejected({ ok: false, msg: "ERROR: Invalid document", errors });
+  };
+
+const doUpsert = (hyper: H, doc: unknown) =>
+  (env: Config) =>
+    Async.of(doc)
+      .chain(ifElse(
+        () => isNil(env.schema),
+        Async.Resolved,
+        validate(prop("schema", env)),
+      ))
+      .chain((doc: unknown) =>
+        compose(hyper.data.get, prop("_id"))(doc)
+          .bichain(
+            () => hyper.data.add(doc),
+            (old: unknown) =>
+              hyper.data.update(prop("_id", old), mergeRight(old, doc)),
+          )
+      );
+
+export const upsert = (hyper: H) =>
+  (id: string, doc: unknown) =>
+    of(mergeRight(doc, { _id: id }))
```

Comment:
This could silently overwrite `_id` in the doc, which may be fine, but a bit of magic. Should this instead check for `_id` on the `doc`, and merge if does not exist, and check for equality if it does exist?

Reply:
The shape of the upsert function is `(id: string, doc: unknown)` and you have to supply the id as an argument, so it should always be the id, even if an id exists on the document.